lib/backend/dynamo/dynamodbbk.go

```diff
@@ -617,45 +657,60 @@ const (
 
 // DeleteRange deletes range of items with keys between startKey and endKey
 func (b *Backend) DeleteRange(ctx context.Context, startKey, endKey backend.Key) error {
-	if startKey.IsZero() {
-		return trace.BadParameter("missing parameter startKey")
-	}
-	if endKey.IsZero() {
-		return trace.BadParameter("missing parameter endKey")
-	}
-	// keep fetching and deleting until no records left,
-	// keep the very large limit, just in case if someone else
-	// keeps adding records
-	for i := 0; i < backend.DefaultRangeLimit/100; i++ {
-		result, err := b.getRecords(ctx, prependPrefix(startKey), prependPrefix(endKey), batchOperationItemsLimit, nil)
+	// Attempt to pull all existing items and delete them in batches
+	// in accordance with the BatchWriteItem limits. There is a hard
+	// cap on the total number of items that can be deleted in a single
+	// DeleteRange call to avoid racing with additional records being added.
+	const maxDeletions = backend.DefaultRangeLimit / 100
+	requests := make([]types.WriteRequest, batchOperationItemsLimit)
+	var pageCount, totalCount int
+	for item, err := range b.Items(ctx, backend.IterateParams{StartKey: startKey, EndKey: endKey}) {
 		if err != nil {
 			return trace.Wrap(err)
 		}
-		if len(result.records) == 0 {
-			return nil
+
+		if totalCount >= maxDeletions {
+			break
 		}
-		requests := make([]types.WriteRequest, 0, len(result.records))
-		for _, record := range result.records {
-			requests = append(requests, types.WriteRequest{
-				DeleteRequest: &types.DeleteRequest{
-					Key: map[string]types.AttributeValue{
-						hashKeyKey:  &types.AttributeValueMemberS{Value: hashKey},
-						fullPathKey: &types.AttributeValueMemberS{Value: record.FullPath},
-					},
+		totalCount++
+
+		requests[pageCount] = types.WriteRequest{
+			DeleteRequest: &types.DeleteRequest{
+				Key: map[string]types.AttributeValue{
+					hashKeyKey:  &types.AttributeValueMemberS{Value: hashKey},
+					fullPathKey: &types.AttributeValueMemberS{Value: prependPrefix(item.Key)},
 				},
-			})
-		}
-		input := dynamodb.BatchWriteItemInput{
-			RequestItems: map[string][]types.WriteRequest{
-				b.TableName: requests,
 			},
 		}
+		pageCount++
+
+		if pageCount == batchOperationItemsLimit {
+			if _, err := b.svc.BatchWriteItem(ctx, &dynamodb.BatchWriteItemInput{
```

Comment:
Does `BatchWriteItem` take ownership of its input or is it only read and then the caller can reuse it afterwards? The previous implementation used a new slice for every batch, which sidesteps the problem.